File: src/main/java/nl/kabisa/util/Result.java
package nl.kabisa.util;

import java.util.function.Function;

public class Result<T, E extends Exception> {
    public static <T, E extends Exception> Result<T, E> ok(T result) {
        return new Result<>(result);
    }

    public static <T, E extends Exception> Result<T, E> err(E exception) {
        return new Result<>(exception);
    }

    public final T result;
    public final E exception;

    private Result(T result) {
        this.result = result;
        this.exception = null;
    }

    private Result(E exception) {
        this.exception = exception;
        this.result = null;
    }

    public boolean isOk() {
        return !isErr();
    }

    public boolean isErr() {
        return exception != null;
    }

    public <U> Result<U, E> map(Function<T, U> mapper) {
        if (isErr())
            return new Result<>(exception);
        else
            return new Result<>(mapper.apply(result));
    }

    public T unwrap() throws E {
        if (isErr())
            throw exception;

        return result;
    }
}

File: src/main/kotlin/nl/kabisa/util/Result.kt
package nl.kabisa.util

import java.util.function.Function

class Result<T, E : Throwable> {

    val result: T?
    val exception: E?

    private constructor(result: T) {
        this.result = result
        this.exception = null
    }

    private constructor(exception: E) {
        this.exception = exception
        this.result = null
    }

    val isOk: Boolean
        get() = !isErr

    val isErr: Boolean
        get() = exception != null

    fun <U> map(mapper: Function<T, U>): Result<U, E> {
        if (isErr)
            return Result(exception!!)
        else
            return Result(mapper.apply(result!!))
    }

    fun unwrap(): T {
        if (isErr)
            throw exception!!

        return result!!
    }

    companion object {
        fun <T, E : Exception> ok(result: T): Result<T, E> {
            return Result(result)
        }

        fun <T, E : Exception> err(exception: E): Result<T, E> {
            return Result(exception)
        }
    }
}
